### PR TITLE
FEATURE: Cross-site linking

### PR DESCRIPTION
**Pull request by aertmann**
Adds ``scheme`` and ``port`` parameters to site domains to enable setting them for cross-site linking.

Additionally adds a primary flag to site domains to enable selecting the primary domain for a site.

Lastly linking to a node using the ``LinkingService`` the primary domain of the site the node belongs
to is taking into account. This allows cross-site linking instead of creating an invalid link to a non-existing
node with the existing site's URL.

NEOS-976 #close
NEOS-178 #close

### CODE REVIEW

**Comment by dlubitz** [comment]
:+1: by reading
Looks great, thanks @aertmann.

File: TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php [hunk 1]
```diff
@@ -37,29 +38,54 @@ class DomainCommandController extends \TYPO3\Flow\Cli\CommandController
...
 
         $domains = $this->domainRepository->findByHostPattern($hostPattern);
         if ($domains->count() > 0) {
-            $this->outputLine('The host pattern "%s" is not unique.', array($hostPattern));
+            $this->outputLine('<error>The host pattern "%s" is not unique.</error>', array($hostPattern));
             $this->quit(1);
         }
 
         $domain = new Domain();
+        if ($scheme) {
```
Comment (kdambekalns) [inline TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php]: `$scheme !== null`, puh-lease :)

File: TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/BackendController.php [hunk 2]
```diff
@@ -64,28 +71,19 @@ public function indexAction()
     /**
      * Redirects to the Neos backend on the given hostname, passing a one-time login token
```
Comment (kdambekalns) [inline TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/BackendController.php]: No more `hostname`, please adjust

File: TYPO3.Neos/Migrations/Mysql/Version20160411101457.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ */
+class Version20160411101457 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     * @return void
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is autogenerated, please modify it to your needs
```
Comment (kdambekalns) [inline TYPO3.Neos/Migrations/Mysql/Version20160411101457.php]: Get rid of this line here and in `down()`, will you? Thanks!

File: TYPO3.Neos/Migrations/Mysql/Version20160411101458.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ */
+class Version20160411101458 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     * @return void
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is autogenerated, please modify it to your needs
```
Comment (kdambekalns) [inline TYPO3.Neos/Migrations/Mysql/Version20160411101458.php]: Get rid of this line here and in `down()`, will you? Thanks!

File: TYPO3.Neos/Migrations/Postgresql/Version20160411101639.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ */
+class Version20160411101639 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     * @return void
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is autogenerated, please modify it to your needs
```
Comment (kdambekalns) [inline TYPO3.Neos/Migrations/Postgresql/Version20160411101639.php]: Get rid of this line here and in `down()`, will you? Thanks!

File: TYPO3.Neos/Migrations/Postgresql/Version20160411101640.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ */
+class Version20160411101640 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     * @return void
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is autogenerated, please modify it to your needs
```
Comment (kdambekalns) [inline TYPO3.Neos/Migrations/Postgresql/Version20160411101640.php]: Get rid of this line here and in `down()`, will you? Thanks!

**Comment by kdambekalns** [comment]
Ok, my usual nitpicks. Otherwise it looks good, didn't test, though.

**Comment by aertmann** [comment]
Rebased

**Comment by aertmann** [comment]
image
image
image
image
image
image
image

**Comment by aertmann** [comment]
Is good to go except for one potential issue. Do we need to add support for setting a path segment for sites running with `local.dev/neos-site/`?

**Comment by aertmann** [comment]
Should be good to go except for one potential issue. Do we need to add support for setting a path segment for sites running with `local.dev/neos-site/`? Keep in mind that it is automatically appended if this existing site uses it, which is probably good enough.

File: TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php [hunk 7]
```diff
@@ -37,29 +38,54 @@ class DomainCommandController extends \TYPO3\Flow\Cli\CommandController
...
+
+        $domainValidator = $this->validatorResolver->getBaseValidatorConjunction(Domain::class);
+        $result = $domainValidator->validate($domain);
+        if ($result->hasErrors()) {
+            foreach ($result->getFlattenedErrors() as $propertyName => $errors) {
+                $firstError = array_pop($errors);
+                $this->outputLine('<error>Validation failed for "' . $propertyName . '": ' . $firstError . '</error>');
+                $this->quit(1);
+            }
+        }
```
Comment (bwaidelich) [inline TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php]: Shouldn't that be validated during persistence anyways?

File: TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php [hunk 8]
```diff
@@ -93,21 +119,37 @@ public function listCommand($hostPattern = null)
...
             ));
             if (strlen($domain->getSite()->getNodeName()) > $longestNodeName) {
                 $longestNodeName = strlen($domain->getSite()->getNodeName());
             }
-            if (strlen($domain->getHostPattern()) > $longestHostPattern) {
-                $longestHostPattern = strlen($domain->getHostPattern());
+            $url =
+                ($domain->getScheme() ? $domain->getScheme() . '://' : '') .
+                $domain->getHostPattern() .
+                ($domain->getPort() ? ':' . $domain->getPort() : '');
```
Comment (bwaidelich) [inline TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php]: What about inlining that "logic" to the Domain class and add some `Domain::getUrl()` method?

File: TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/MenuHelper.php [hunk 9]
```diff
@@ -64,8 +64,10 @@ public function buildSiteList(ControllerContext $controllerContext)
             $active = false;
             /** @var $site \TYPO3\Neos\Domain\Model\Site */
             if ($site->hasActiveDomains()) {
-                $hostname = $site->getFirstActiveDomain()->getHostPattern();
-                $active = $hostname === $requestUriHost;
+                $activeDomains = $site->getActiveDomains()->map(function ($domain) {
```
Comment (bwaidelich) [inline TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/MenuHelper.php]: <nitpick>wouldn't the result of this map be a list of `HostPattern`-strings? So maybe call this `$activeHostPatterns` instead?</nitpick>

File: TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php (hunk 7 above)
Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php]: that would be nice, but wouldn't add it if it did would I? ;) guess it's not handled automatically for commands

**Comment by bwaidelich** [comment]
@aertmann thanks so much for taking care of this! I didn't test yet, but the code looks very nice. I just added the usual nitpick comments (because I couldn't find anything severe ;)
Regarding your question of allowing sites to specify some path prefix (if that's what you meant?) – I'm not sure but my gut feeling is: If you're not able to use virtual hosts for some reason you probably don't need the feature of cross-site linking anyways.. At least I tend to say that we should only allow this if people really need it to prevent unexpected behavior

File: TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php (hunk 8 above)
Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php]: it is and can be used here, but not further down.. mixed it up at some point

File: TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/MenuHelper.php (hunk 9 above)
Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/MenuHelper.php]: nope, this returns the host from the domain instead of the domains..

**Comment by aertmann** [comment]
fixed nipicks

File: TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php (hunk 7 above)
Comment (bwaidelich) [inline TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php]: I just tested it (without your patch) and get:
image

File: TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/MenuHelper.php (hunk 9 above)
Comment (bwaidelich) [inline TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/MenuHelper.php]: What I meant is, that after this call `$activeDomains` is an array of strings representing host patterns of active domains, so the name `$activeDomain` is a bit misleading.. Not a biggy though

Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Controller/Backend/MenuHelper.php]: good point, fixed

File: TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php (hunk 7 above)
Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Command/DomainCommandController.php]: Hmm interesting. I get this:

Without:
image
With:
image

**Comment by aertmann** [comment]
should be merge ready, anybody up for a proper review? 😉

**Comment by aertmann** [comment]
rebased (fixed merge conflict)

**Comment by kitsunet** [comment]
Thanks for taking care, voted and will merge in 24 if no one objects.

**Comment by aertmann** [comment]
thanks for the reviews 👍

**Pull request merged**